WordPress/gutenberg
# Scripts: Include variations paths in build

## What?
If a `block.json`'s `variations` field is set to a (PHP) file name, include that file in the build, i.e. prefix its functions with `gutenberg_` and copy the resulting file to the `build/block-library/blocks/` directory.

Note that this applies to third-party blocks, but not Core blocks. Making it work for Core blocks is a separate issue: #63077.

## Why?
As of #62092, it is possible to set the `variations` field in a block's `block.json` file to the name of a PHP file that generates the block's variations. 

## How?
By extending the existing logic introduced for the `render` field (by #43917 and #51162) to encompass `variations`.

## Testing Instructions

- On this branch, apply the patch below to create a new subdirectory that contains a basic example block that has a `block.json` file with a `variations` field that points to a PHP file.
- `cd example-variations/`
- Run `../node_modules/.bin/wp-scripts build`.
- Verify that `src/variations.php` has been copied to the `build/` folder.
- Set the `variations` field in `src/block.json` to a variations array: `"variations": [ { "name": "warning" } ]`.
- Run `../node_modules/.bin/wp-scripts start` to begin development in watch mode.
- There should be _no_ `variations.php` file in the `build/` folder.
- Set the `variations` field in `src/block.json` back to `"variations": "file:./variations.php"`.
- The `variations.php` file should re-appear in the `build/` folder.

<details>
<summary>
Patch
</summary>

```diff
diff --git a/example-variations/package.json b/example-variations/package.json
new file mode 100644
index 00000000000..5a39f05d9ab
--- /dev/null
+++ b/example-variations/package.json
@@ -0,0 +1,8 @@
+{
+	"name": "example-variations",
+	"version": "0.1.0",
+	"description": "Example block scaffolded with Create Block tool.",
+	"author": "The WordPress Contributors",
+	"license": "GPL-2.0-or-later",
+	"scripts": {}
+}
diff --git a/example-variations/src/block.json b/example-variations/src/block.json
new file mode 100644
index 00000000000..200e3fa493f
--- /dev/null
+++ b/example-variations/src/block.json
@@ -0,0 +1,5 @@
+{
+	"name": "example/variations",
+	"title": "Block with Variations",
+	"variations": "file:./variations.php"
+}
diff --git a/example-variations/src/variations.php b/example-variations/src/variations.php
new file mode 100644
index 00000000000..bed66d95441
--- /dev/null
+++ b/example-variations/src/variations.php
@@ -0,0 +1,10 @@
+<?php
+
+return array(
+	array(
+		'name'        => 'warning',
+		'title'       => 'warning',
+		'description' => 'Shows warning.',
+		'keywords'    => array( 'warning' ),
+	),
+);
```
</details
